dimod/include/dimod/adjvectorbqm.h

```diff
@@ -83,6 +84,96 @@ class AdjVectorBQM {
         }
     }
 
+    /**
+     * Construct a BQM from a dense array. This constructor is parallelized
+     * and temporarily zeroes out the diagonal of the dense array but restores
+     * it back.
+     *
+     * @param dense An array containing the biases. Assumed to contain
+     *     `num_variables`^2 elements. The upper and lower triangle are summed.
+     * @param num_variables The number of variables.
+     */
+    template <class B2>
+    AdjVectorBQM(B2 dense[], size_type num_variables,
```

Comment:
To avoid check in the innermost loop I need to zero out the diagonal entries of the source matrix temporarily and restore it back. I restore it back to maintain consistency with the existing constructor that uses a read-only dense matrix. Making it read-write can only lead to an issue when multiple threads are accessing the same matrix in different functions. In this function I access it using multiple threads but I handle the synchronization myself so it is okay.